### ISSUE TITLE
Projects using Rhai

### ISSUE THREAD

**Issue by stevedonovan** [rhaiscript/rhai#69]
Just a tracking issue for projects.

Apparently I've got the first published crate that uses Rhai as a dependency:

https://github.com/stevedonovan/findr

**Comment by luciusmagn** [comment]
Awesome! I like the idea of a find that doesn't have an arcane obscure syntax. Yep, you are the first person to actually have a crate using it published. Good job :D

**Comment by juanibiapina** [comment]
I'm starting this experiment with rhai: https://github.com/juanibiapina/roadrunner/pull/16

I'll keep improving it in the next few days and decide if I'll move forward with it or go back to my custom markup language.

**Comment by juanibiapina** [comment]
I've decided to use Lua instead.

**Comment by zesterer** [comment]
@luciusmagn We're considering using Rhai for a game project (https://www.github.com/veloren/game/). The team has a few questions about the language first though. Where is best to ask them?

**Comment by luciusmagn** [comment]
I'd say here, or you can write me an email at luk.hozda@gmail.com. I am
currently out of home, so it may take a while for me to respond

Dne pá 6. 7. 2018 10:48 uživatel Joshua Barretto <notifications@github.com>
napsal:

**Comment by GopherJ** [comment]
[casbin-rs](https://github.com/casbin/casbin-rs) is using rhai

**Comment by MarkuBu** [comment]
I consider using Rhai for a game too. But I'm in the planning process and far away from any code

Another option for scripting would be Mun, but this is also in an early state

**Comment by schungx** [comment]
@MarkuBu   my $0.02 worth - use Rhai if you want tight integration with lots of functionalities written in Rust.  If you want to basically write your game logic in a scripting language, probably Lua is much more suited for this purpose.

**Comment by MarkuBu** [comment]
@schungx why? Because of the performance? I know that Lua is fast, even without the JIT

**Comment by schungx** [comment]
Yes.  Rhai is an AST-walking interpreter.  It is never going to break any speed records.  So use it where the majority of your application's time is not spend in the script.

**Comment by MarkuBu** [comment]
Ok, thanks.

**Comment by sunng87** [comment]
I'm looking for a scripting language to add dynamic helper support for my template engine [handlebars-rust](https://github.com/sunng87/handlebars-rust). By far Rhai seems to be a good fit because of its inter-operatability with rust data. I will be writing poc code recently and seeking for help with you.

**Comment by schungx** [comment]
No prob!

**Comment by sunng87** [comment]
FYI support for rhai script helper has been released in handlebars 3.1.0 with a feature flag `script_helper` https://github.com/sunng87/handlebars-rust/blob/master/examples/script.rs

**Comment by Pebaz** [comment]
I'm trying to integrate Rhai into my game butbfor the life of me I can't figure out how to pass a Vec of a custom Rust type (already registered with the engine) and access the indices from it.

I can successfully pass it but when the script tries to access any index it crashes.

**Comment by schungx** [comment]
When you pass a `Vec` into Rhai, it treats it as a custom type.  You're responsible for registering an API for use with that API.  In your case, you'd want to register an _indexer_.

```rust
engine
    .register_type_with_name::<Vec<MyType>>("Vec<MyType>")
    .register_indexer_get_set(
        |v: &mut Vec<MyType>, i: INT| v[i],
        |v: &mut Vec<MyType>, i: INT, value: MyType| v[i] = value)
    .register_fn("len", |v: &mut Vec<MyType>| v.len() as INT)
    .register_get("len", |v: &mut Vec<MyType>| v.len() as INT)
    .register_fn("contains", |v: &mut Vec<MyType>, value: MyType| v.contains(&value))
    .register_fn("is_empty", |v: &mut Vec<MyType>| v.is_empty())
    .register_get("is_empty", |v: &mut Vec<MyType>| v.is_empty());
```

**Comment by schungx** [comment]
Alternatively, you can convert the `Vec<T>` into a Rhai `Array` and it'll work with Rhai just fine.

```rust
// assuming 'my_vec' is 'Vec<T>'
let array: Dynamic = my_vec.into();
```

**Comment by Pebaz** [comment]
@schungx Thank you very much for your help this is what I needed!

**Comment by koute** [comment]
I'm using Rhai as a DSL for my [memory profiler](https://github.com/koute/memory-profiler).

**Comment by garypen** [comment]
We are using rhai in our new GraphQL router project: https://github.com/apollographql/router 

The router has a "plugin" framework which allows customer to write plugins in rust which may inspect or modify the flow of data through the router. We have a rhai plugin, which exposes router functionality so that plugins can also be written in rhai. This is nice for our customers, since it's much simpler to write a rhai plugin than a rust plugin.

Support is deemed to be experimental at the moment, but our experience so far is good!

**Comment by schungx** [comment]
This is interesting.  Is your payload mostly binary-encoded or text-encoded (like JSON)?

**Comment by garypen** [comment]
It's text-encoded (JSON), so it's mainly about manipulating pre-filled data objects.

Your comment made me chuckle because I originally (before I found out about Fn()) was just calling functions by string representation. I can't make my mind up if that looks better or not, but I'll try it out in a few examples so people are aware it exists.

We are using closures in one of the examples in a soon to land PR...

**Comment by schungx** [comment]
> Your comment made me chuckle because I originally (before I found out about Fn()) was just calling functions by string representation. I can't make my mind up if that looks better or not, but I'll try it out in a few examples so people are aware it exists.

Why not both?  I find Rhai's function overloading to be very useful in this regard, almost like writing JavaScript.

But you have to physically overload the function with a version that takes `&str` parameter in order for that to work, so it is essentially a duplicated API.

**Comment by ltabis** [comment]
We are using rhai in [vSMTP](https://github.com/viridIT/vSMTP), a mail transfer agent made with rust.

Instead of filtering emails with a standard configuration format (like json or toml) we use Rhai to make powerful rules for filtering, enabling you to execute behavior on specific email addresses, clients, ip addresses etc ...

We are constantly trying to optimize the "rule engine" (the part that handles rhai) because the server needs to be really fast. But with the help of the team & rhai's great docs we are getting there :)

**Comment by schungx** [comment]
BTW, just another tip-up.  In your docs, you can link to https://rhai.rs/book/ref for a "Language Reference" section that contains only the scripting language, not the Rust engine bits.  It would be easier on your users.

**Comment by YaLTeR** [comment]
I'm using Rhai in https://github.com/YaLTeR/bxt-rs to script the goal and constraints for a brute-force optimizer. At first it was Lua (luajit), but then I couldn't figure out how to make it cross-compile out of the box on all system configurations that I care about, and tried Rhai as a pure-Rust option. Seems to work great so far; on the small scripts that I'm using it doesn't seem any slower than luajit.

**Comment by schungx** [comment]
> it doesn't seem any slower than luajit.

If you're not repeatedly running a script in a tight loop, then the JIT doesn't really show its power...

**Comment by YaLTeR** [comment]
During the optimization process normally the rest of the code takes much longer than the goal checking script, so yeah.

One performance bottleneck that I did hit with both luajit and Rhai is passing large arrays to the script (I guess specifically converting to luajit's tables or Rhai's Dynamic type). The script accepts either a single struct with a few fields (player data for the last state) or an array with several hundred of those structs (player data for all states). Ideally I'd unconditionally pass all states but it's quite slow to convert. Not sure how to improve this in a way that doesn't require writing manual field access or serialization code.

**Comment by schungx** [comment]
An option is to wrap that data structure in an `Rc<RwLock<Vec<Rc<RwLock<Item>>>>>`.  Then it is shared access.

But then you'll have to register the relevant API's to access elements in (e.g. indexers) and fields access for `Item`.

But for high performance work when you don't want to copy things, it is worth it.

**Comment by erlend-sh** [comment]
Other found via [Used by](https://github.com/rhaiscript/rhai/network/dependents):

- https://github.com/ballerine-io/ballerine
- https://github.com/rustic-rs/rustic
- https://github.com/rabbit-digger/rabbit-digger-pro
- https://github.com/markusmoenig/ForgedThoughts
- https://github.com/PikuseruConsole/pikuseru
- https://github.com/Notgnoshi/generative
- https://github.com/BrettMayson/HEMTT
- https://github.com/mkeeter/fidget
- https://github.com/phimuemue/openschafkopf
- https://github.com/FyraLabs/anda
- https://github.com/nervosnetwork/ckb
-

**Comment by ryankopf** [comment]
I am using RHAI in https://rpgfx.com/ RPG Studio

**Comment by elenapan** [comment]
This dotfile manager uses Rhai for configuration purposes:
https://github.com/elkowar/yolk/
